gto/git_utils.py

```diff
@@ -0,0 +1,71 @@
+import inspect
+import logging
+import re
+from functools import wraps
+from tempfile import TemporaryDirectory
+from typing import Callable, Dict
+
+from git import Repo
+
+# TODO: make a new type out of Union[str, Repo]
+
+
+def git_clone_if_repo_is_remote(f: Callable):
+    @wraps(f)
+    def wrapped_f(*args, **kwargs):
+        kwargs = _turn_args_into_kwargs(args, kwargs)
+
+        # noinspection PyTypeChecker
+        if isinstance(kwargs["repo"], str) and is_url_to_remote_repo(
+            repo=kwargs["repo"]
+        ):
+            try:
+                with TemporaryDirectory() as tmp_dir:
+                    logging.debug("create temporary directory %s", tmp_dir)
+                    # noinspection PyTypeChecker
+                    git_clone(repo=kwargs["repo"], dir=tmp_dir)
+                    kwargs["repo"] = tmp_dir
+                    result = f(**kwargs)
+            except (NotADirectoryError, PermissionError) as e:
+                raise e.__class__(
+                    "Are you using windows with python < 3.9? "
+                    "This may be the reason of this error: https://bugs.python.org/issue42796. "
+                    "Consider upgrading python."
+                ) from e
+            logging.debug("temporary directory %s has been deleted", tmp_dir)
+        else:
+            result = f(**kwargs)
+
+        return result
+
+    def _turn_args_into_kwargs(
+        args: tuple, kwargs: Dict[str, object]
+    ) -> Dict[str, object]:
+        kwargs_complement = {
+            k: args[i]
+            for i, k in enumerate(inspect.getfullargspec(f).args)
+            if k not in kwargs.keys() and i < len(args)
+        }
+        kwargs.update(kwargs_complement)
+        return kwargs
+
+    return wrapped_f
+
+
+def git_clone(repo: str, dir: str) -> None:
+    logging.debug("clone %s in directory %s", repo, dir)
+    Repo.clone_from(url=repo, to_path=dir)
```

Comment:
I did it in this way at first :) but then realized: we are now writing some interface function that should eventually work even if change git tool. Currently we are using `gitpython` but let's say that we then will need to switch. Perhaps there won't be any other choice but to call `git` from the shell. Then it is better to keep the `git_clone` function interface not returning anything (could return the path to the git dir, but that's an input given, it does not make sense to return it).

One additional argument. As far as I know, it is good practice to separate queries from commands. Commands change the state of the machine, and should not return anything. And `git_clone` is a command...

Reply:
Ok, got your argument. Since it's internals not exposed to users, I think we could do the either way (return something that can change or not return it), so it's up to you.

The other question - maybe move `with TemporaryDirectory() as tmp_dir` here and make this function itself a context manager? Like this https://github.com/iterative/gto/blob/0293e674593293f76ee5ad288dcb5101edc8415f/tests/git_utils/test_git_clone_if_repo_is_remote.py#L63

Then you'll split the cloning logic from the wrapper logic.

Reply:
I was thinking about something like [this](https://www.geeksforgeeks.org/context-manager-using-contextmanager-decorator/) to isolate temp dir handling from the wrapper itself:
```python
from contextlib import contextmanager

@contextmanager
def git_clone(repo: str) -> str:
    tmp_dir = TemporaryDirectory()
    logging.debug("clone %s in directory %s", repo, tmp_dir)
    Repo.clone_from(url=repo, to_path=tmp_dir)
    yield tmp_dir
    tmp_dir.cleanup()
```

Reply:
Then it would be
```python
def git_clone_if_repo_is_remote(f: Callable):
    ...
            try:
                with git_clone(repo=repo) as tmp_dir:
                    return f(tmp_dir, *args, **kwargs)
            except (NotADirectoryError, PermissionError) as e:
```

Reply:
This is basically the suggestion you have already done here: https://github.com/iterative/gto/issues/25#issuecomment-1246585717

But, as said, I strongly advise against using a `yield` statement: https://github.com/iterative/gto/issues/25#issuecomment-1246700348

Reply:
Hmm, maybe I misunderstand something? This is different, since in fact you just write a context manager and work with that context manager as you should. Note `with git_clone(repo=repo) as tmp_dir` - when `f` function runs and returns the result, the control will be given to the context manager and `tmp_dir.cleanup()` will be called.

Reply:
sorry, my bad! Too much multi-tasking.

Ok, will do a change along these lines.

Reply:
Done, I introduced `cloned_git_repo`.

Still, I don't want to give up on the decorator able to handle a function where `repo` is not necessarily the first argument :P

Reply:
Yes, let's keep it, no problem)